### DISCUSSION TITLE
9.x.x, Webpack, and Service Workers

### DISCUSSION THREAD

**Discussion by eliw00d** [firebase/firebase-js-sdk#5730] [Q&A] [answered]
I am having a heck of a time finding documentation anywhere for how to go from this:
```javascript
// Give the service worker access to Firebase Messaging.
// Note that you can only use Firebase Messaging here. Other Firebase libraries
// are not available in the service worker.
importScripts('https://www.gstatic.com/firebasejs/8.10.0/firebase-app.js');
importScripts('https://www.gstatic.com/firebasejs/8.10.0/firebase-messaging.js');

// Initialize the Firebase app in the service worker by passing in
// your app's Firebase config object.
// https://firebase.google.com/docs/web/setup#config-object
firebase.initializeApp({
  apiKey: 'api-key',
  authDomain: 'project-id.firebaseapp.com',
  databaseURL: 'https://project-id.firebaseio.com',
  projectId: 'project-id',
  storageBucket: 'project-id.appspot.com',
  messagingSenderId: 'sender-id',
  appId: 'app-id',
  measurementId: 'G-measurement-id',
});

// Retrieve an instance of Firebase Messaging so that it can handle background
// messages.
const messaging = firebase.messaging();
```

to this:
```javascript
import { initializeApp } from "firebase/app";
import { getMessaging } from "firebase/messaging/sw";

// Initialize the Firebase app in the service worker by passing in
// your app's Firebase config object.
// https://firebase.google.com/docs/web/setup#config-object
const firebaseApp = initializeApp({
  apiKey: 'api-key',
  authDomain: 'project-id.firebaseapp.com',
  databaseURL: 'https://project-id.firebaseio.com',
  projectId: 'project-id',
  storageBucket: 'project-id.appspot.com',
  messagingSenderId: 'sender-id',
  appId: 'app-id',
  measurementId: 'G-measurement-id',
});

// Retrieve an instance of Firebase Messaging so that it can handle background
// messages.
const messaging = getMessaging(firebaseApp);
```

I am guessing I'll need to register the service worker with type module, but how do I support that with Webpack? Currently, I am using the 'copy-webpack-plugin' to copy the service worker file to the public path. That works, but now I need to support the import syntax, and scouring the web has not produced many results. The documentation basically just says "we assume you know how to do this". Unfortunately, even Webpack's own documentation is lacking, so it would be nice to have an example to go by. Any help?

**Comment by adamhari** [comment, marked as answer]
Here's what I did to build the firebase-messaging-sw.js file with webpack (I'm not sure if this works because I'm having other issues with Firebase, I would be interested in knowing):

1. Add an additional item in the config's `entry` collection for the firebase-messaging-sw.js file. We do this because we want to bundle and output it separately from the rest of the app's files.

```
entry: {
  ...,
  'firebase-messaging-sw': path.join(__dirname, './web/firebase-messaging-sw.ts'),
}
```

2. Change your `output.filename` config if needed so that the service worker entry's output file is named correctly. Mine is doing something like this:

```
pathData => {
  return pathData.chunk.name === 'firebase-messaging-sw' ? '[name].js' : '[name].[hash].js';
},
```

3. Add the service worker file to your HtmlWebpackPlugin's `excludeChunks` collection. We do this because we don't want the script tag to be included and executed from the outputted index.html file, we only want it to exist in the outputted directory.

4. Disable `optimization.runtimeChunk`. I didn't want to do this, but it was affecting the naming and content of the service worker file.

**Reply by eliw00d** [reply]
I actually ended up figuring this out after a lot of time reading the webpack docs and ended up with something nearly identical to your solution. I totally forgot to follow up until today, so thank you for taking the time and providing an answer for others to see as well!

I didn't have to do anything like steps 3 or 4, though.

Any thoughts on how to replace service workers for new deployments? Since switching to register type module, it seems like previous non-module SWs just get stuck in a stopped state.

**Reply by adamhari** [reply]
@eliw00d Not sure about that one! I figure step 3 and 4 might have only been necessary for me because I'm overriding the defaults from the webpack preset I use (@expo/webpack-config).

**Reply by eliw00d** [reply]
@adamhari 
~Turns out step 3 is exactly what I needed to resolve: https://github.com/firebase/firebase-js-sdk/discussions/6025.~ Nevermind, it still happens.

I am using Webpack v5 and ran across https://github.com/webpack/webpack/issues/6598#issuecomment-770698232 which says you shouldn't need to disable `optimization.runtimeChunk` anymore, if I understand correctly.

**Reply by adamhari** [reply]
@eliw00d I'm not so sure. With `runtimeChunk` enabled it was splitting the compiled runtime/non-runtime contents of `firebase-messaging-sw.ts` into separate files in addition to changing the filenames. I suspect that the service worker needs all the contents in the same file with the correct name.

**Reply by eliw00d** [reply]
@adamhari 
Are you able to share what it looks like in `sources` in Chrome? 

Mine used to be in `top/<website>/firebase-messaging-sw.js`, like this:

![image](https://user-images.githubusercontent.com/1801127/155580227-fa1c6983-3c0f-4016-9484-839a9bb7a9c2.png)

but is now just in `firebase-messaging-sw.js` and has the gear icon, like this:

![image](https://user-images.githubusercontent.com/1801127/155580066-b422c2db-1b19-4986-aacb-c73b8c4c6830.png)

The only split I can see inside `firebase-messaging-sw.js` is between `<website>` and a `webpack://` entry: 

![image](https://user-images.githubusercontent.com/1801127/155580448-c594e363-6c99-43df-91da-f7c36dc8641d.png)

**Reply by adamhari** [reply]
![image](https://user-images.githubusercontent.com/16565387/155772257-14129df9-2528-4ef8-abcd-4c3bbaa3cce8.png)

**Reply by eliw00d** [reply]
I guess I'm not seeing the split you're referring to, mine looks pretty similar to yours, the only difference being my `.js` file is minimized/transpiled and I don't have a `.ts` file.

**Reply by adamhari** [reply]
Mine is only split when i have `runtimeChunk` enabled, that screenshot is with my current config where I don't have it enabled.

**Reply by tyuyucu** [reply]
Thank you @adamhari & @eliw00d for opening the issue
All I needed were Steps 1 & 3 to achieve the same thing under Vue's and some other step which you will only have to face if you use vue

My `vue.config.js` file:
```js
  devServer: {
  configureWebpack: config => (
    (config.entry["firebase-messaging-sw"] = ["./src/firebase-messaging-sw.js"]),undefined),
  chainWebpack: config => {
    config.plugin('html').tap(args => ((args[0].excludeChunks = ['firebase-messaging-sw']),args))
  }
};
```

And here is the extra step for vue, which I had to do use, since Vue configures webpacks output to be inside a /js directory and firebase by default needs the sw file to be at root location except configured otherwise explicitly.

My `initFCM.js` file:
```js
  const serviceWorkerRegistration = await navigator.serviceWorker.register(
    "/js/firebase-messaging-sw.js"
  );
  const token = await getToken(messaging, {
    vapidKey: "[insertYourOwnVapidKeyHere]",
    serviceWorkerRegistration,
  });
```

Firebase how about adding those steps in the doc for all webpack users?
I already sent Firebase a feedback on their doc page which should actually explain those steps or atleast link to another doc page which does. I also included this issue in the feedback so all they gotta do is copy paste customize.

**Reply by eliw00d** [reply]
@adamhari With type module did you end up having to change updateViaCache to `'none'` or anything like that? Documentation is scarce but [I can't seem to figure out why my service worker stops after some period of time](https://github.com/firebase/firebase-js-sdk/discussions/6025). I feel like with `type: 'classic'` it never had issues. Just curious if you've had similar issues.

**Reply by adamhari** [reply]
@eliw00d I don't ever call `navigator.serviceWorker.register()` directly.

**Reply by eliw00d** [reply]
> @eliw00d I don't ever call `navigator.serviceWorker.register()` directly.

Ah okay, the reason I ended up doing so (but copied the way the SDK does it) was to pass URL search params for environment variables. I will try to find a way to avoid that.

**Reply by eliw00d** [reply]
Do you do anything differently for localhost?

**Reply by adamhari** [reply]
No, other than env vars (unique Firebase project for each env) everything is handled the same.

**Reply by eliw00d** [reply]
Sounds good. Must be something specific to my use of devServer, then. If I build the production version locally it works fine.
